gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/TXState.java

```diff
@@ -1159,6 +1160,49 @@ public boolean execute(Object key) {
     return eventsToFree;
   }
 
+  private void publishRecordedVersions() {
+    final TXRegionState[] finalRegions = this.finalizeRegions;
+    GemFireCacheImpl cache = GemFireCacheImpl.getExisting();
+    final LogWriterI18n logger = getTxMgr().getLogger();
+    // No need to check for snapshot if we want to enable it for RC.
+    if (cache.snapshotEnabled()) {
+      if (isSnapshot() || cache.snapshotEnabledForTX()) {
+        // first take a lock at cache level so that we don't go into deadlock or sort array before
+        // This is for tx RC, for snapshot just record all the versions from the queue
+        cache.acquireWriteLockOnSnapshotRvv();
```

Comment:
It is better to acquire the lock inside try. Here it does not matter but if someone refactors a bit and there is a scope of failure between acquire and try block start then the write lock won't be released.